os/arch/arm/src/amebad/amebad_i2s.c

```diff
@@ -0,0 +1,1818 @@
+
+static uint32_t i2s_txdatawidth(struct i2s_dev_s *dev, int bits)
+{
+#if defined(I2S_HAVE_TX) && (0 < I2S_HAVE_TX)
+	struct amebad_i2s_s *priv = (struct amebad_i2s_s *)dev;
+
+	/* Support 16, 24, 32 bits */
+	DEBUGASSERT(priv && (bits == I2S_BITS_PER_SAMPLE_16BIT /
+				|| bits == I2S_BITS_PER_SAMPLE_32BIT /
+				|| bits == I2S_BITS_PER_SAMPLE_24BIT));
+
+	priv->bits_per_sample = bits;
+
+	/* amebad 16, 24, 32, bits setting */
+	if (bits == I2S_BITS_PER_SAMPLE_16BIT) priv->i2s_object.word_length = WL_16b;
+	else if (bits == I2S_BITS_PER_SAMPLE_24BIT) priv->i2s_object.word_length = WL_24b;
+	else if (bits == I2S_BITS_PER_SAMPLE_32BIT) priv->i2s_object.word_length = WL_32b;
+
+	return priv->bits_per_sample * priv->sample_rate;
+#endif
+
+	return 0;
+}
+
+/****************************************************************************
+* Name: i2s_send
+*
+* Description:
+*   Send a block of data on I2S.
+*
+* Input Parameters:
+*   dev      - Device-specific state data
+*   apb      - A pointer to the audio buffer from which to send data
+*   callback - A user provided callback function that will be called at
+*              the completion of the transfer.  The callback will be
+*              performed in the context of the worker thread.
+*   arg      - An opaque argument that will be provided to the callback
+*              when the transfer complete
+*   timeout  - The timeout value to use.  The transfer will be canceled
+*              and an ETIMEDOUT error will be reported if this timeout
+*              elapsed without completion of the DMA transfer.  Units
+*              are system clock ticks.  Zero means no timeout.
+*
+* Returned Value:
+*   OK on success; a negated errno value on failure.  NOTE:  This function
+*   only enqueues the transfer and returns immediately.  Success here only
+*   means that the transfer was enqueued correctly.
+*
+*   When the transfer is complete, a 'result' value will be provided as
+*   an argument to the callback function that will indicate if the transfer
+*   failed.
+*
+****************************************************************************/
+
+static int i2s_send(struct i2s_dev_s *dev, struct ap_buffer_s *apb, i2s_callback_t callback, void *arg, uint32_t timeout)
+{
+	struct amebad_i2s_s *priv = (struct amebad_i2s_s *)dev;
+
+	DEBUGASSERT(priv && apb);
+
+	i2sinfo("[I2S TX] apb=%p nbytes=%d arg=%p samp=%p timeout=%d\n", apb, apb->nbytes - apb->curbyte, apb->samp, arg, timeout);
+
+	i2s_dump_buffer("Sending", &apb->samp[apb->curbyte], apb->nbytes - apb->curbyte);
+
+#if defined(I2S_HAVE_TX) && (0 < I2S_HAVE_TX)
+	struct amebad_buffer_s *bfcontainer;
+	irqstate_t flags;
+	int ret;
+
+	/* Has the TX channel been enabled? */
+	if (!priv->txenab) {
+		i2serr("ERROR: I2S has no transmitter\n");
+		return -EAGAIN;
+	}
+
+	/* Allocate a buffer container in advance */
+
+	bfcontainer = i2s_buf_tx_allocate(priv);
+	DEBUGASSERT(bfcontainer);
```

Comment:
Is this needed? Because in `i2s_buf_tx_allocate()`, it makes assert if `bfcontainer` is NULL.

Reply:
Removed `DEBUGASSERT(bfcontainer);`, as there is ASSERT for `bfcontainer` inside `i2s_buf_tx_allocate()`.